### PR TITLE
Groups, Group Members, Assignments API

### PR DESCRIPTION
**Pull request by Nitish145**
#### Describe the changes you have made in this PR -
1. Adds the following Groups endpoints
    - `/api/v1/groups` to fetch all groups.
    - `/api/v1/groups_mentored` to fetch the groups you mentor
    - `/api/v1/groups/` to POST a new group.
    - `/api/v1/groups/:id/` to GET, UPDATE, DELETE group details.

2. Adds the following Group Members endpoints
    - `/api/v1/groups/:group_id/group_members` to GET all the group members or POST new group member.
    - `/api/v1/group_members/:id/` to DELETE any group member.

3. Adds the following Assignments endpoints
    - `/api/v1/groups/:group_id/assignments` to GET all assignments for a group or POST new assignment
    - `/api/v1/assignments/:id` to GET assignment details, PATCH or DELETE assignment.
    - `/api/v1/assignments/:id/reopen` to reopen a closed assignment.
    - `/api/v1/assignments/:id/start` to start working on any assignment

4. Add Groups, Group Members, Assignments schema

5. Add specs for each of the possible test cases.

### The updated `docs` are available [here](https://nitish145.github.io/slate/)

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by Nitish145** [comment]
@tachyons please check!!

File: app/controllers/api/v1/assignments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    end
+
+    def assignment_update_params
+      params.require(:assignment).permit(
+        :name, :deadline, :description, :restrictions
+      )
+    end
+
+    def check_access
+      authorize @assignment, :admin_access?
```
Comment (tachyons) [inline app/controllers/api/v1/assignments_controller.rb:104]: Why admin access is needed to create an assignment ?

File: app/controllers/api/v1/group_members_controller.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+class Api::V1::GroupMembersController < Api::V1::BaseController
+  before_action :authenticate_user!
+  before_action :set_group, only: %i[index create]
+  before_action :check_edit_access, only: %i[create]
+
+  # GET /api/v1/groups/:group_id/group_members/
+  def index
+    raise ActiveRecord::RecordNotFound if @group.group_members.empty?
```
Comment (tachyons) [inline app/controllers/api/v1/group_members_controller.rb:10]: This should return an empty array, not 404

File: app/controllers/api/v1/assignments_controller.rb (hunk 1 above)
Comment (Nitish145) [reply app/controllers/api/v1/assignments_controller.rb:104]:
```ruby
def admin_access?
  (assignment.group&.mentor_id == user.id) || user.admin?
end
```
Group mentor or some admin can only create assignments right ? @tachyons

Comment (tachyons) [reply app/controllers/api/v1/assignments_controller.rb:104]: That name is weird, ideally that name should be update_access or some thing similar

File: app/controllers/api/v1/group_members_controller.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        GroupMember.where(group_id: @group.id, user_id: user.id).first_or_create
+      end
+    end
+
+    render json: { added: @added_mails, pending: @pending_mails, invalid: @invalid_mails }
+  end
+
+  # DELETE /api/v1/group_members/:id
+  def destroy
+    @group_member = GroupMember.find(params[:id])
```
Comment (tachyons) [inline app/controllers/api/v1/group_members_controller.rb:42]: Why we aren't using beore filter to set GroupMember here ?

Comment (Nitish145) [reply app/controllers/api/v1/group_members_controller.rb:42]: Since it was just one time use, i didn't used `before_action` but we surely can..

Comment (Nitish145) [reply app/controllers/api/v1/group_members_controller.rb:42]: Since it was just one time use, i didn't used `before_action` but we surely can.. @tachyons

Comment (Nitish145) [reply app/controllers/api/v1/group_members_controller.rb:42]: Done!

File: app/controllers/api/v1/assignments_controller.rb (hunk 1 above)
Comment (Nitish145) [reply app/controllers/api/v1/assignments_controller.rb:104]: Can this be updated in this PR only ? @tachyons

**Comment by Nitish145** [comment]
@tachyons please check..

**Comment by Nitish145** [comment]
@tachyons What further changes could be done here ?

File: app/controllers/api/v1/assignments_controller.rb (hunk 1 above)
Comment (Nitish145) [reply app/controllers/api/v1/assignments_controller.rb:107]: @tachyons any updates here ?

Comment (tachyons) [reply app/controllers/api/v1/assignments_controller.rb:107]: No, out of the scope for this PR

File: app/controllers/api/v1/group_members_controller/mails_handler.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+class Api::V1::GroupMembersController
+  class MailsHandler
```
Comment (tachyons) [inline app/controllers/api/v1/group_members_controller/mails_handler.rb:4]: Keep this in app/services

File: app/serializers/api/v1/assignment_serializer.rb [hunk 5]
```diff
@@ -9,5 +9,11 @@ class Api::V1::AssignmentSerializer
     (assignment.group.mentor_id == params[:current_user].id) || params[:current_user].admin?
   end
 
+  attributes :your_project_id do |assignment, params|
```
Comment (tachyons) [inline app/serializers/api/v1/assignment_serializer.rb:12]: This is a leaky abstraction, add the attribute `project_author_id` and find own project in client side or create a separate endpoint for this

File: config/routes.rb [hunk 6]
```diff
@@ -136,6 +136,18 @@
...
       resources :grades, only: [:update, :destroy]
+      resources :groups, only: [:index, :show, :update, :destroy]
+      get '/groups_mentored', to: 'groups#groups_mentored'
+      resources :groups do
+        resources :group_members, shallow: true
+        resources :assignments, shallow: true
+      end
+      resources :assignments do
+        member do
+          get 'reopen'
```
Comment (tachyons) [inline config/routes.rb:147]: GET requests are for idempotent end points, this should be PUT/PATCH

File: config/routes.rb [hunk 7]
```diff
@@ -136,6 +136,18 @@
...
+          get 'start'
```
Comment (tachyons) [inline config/routes.rb:148]: same as above

File: app/serializers/api/v1/group_serializer.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+class Api::V1::GroupSerializer
+  include FastJsonapi::ObjectSerializer
+
+  attributes :member_count do |group|
+    group.group_members.count
```
Comment (tachyons) [inline app/serializers/api/v1/group_serializer.rb:7]: can we use counter cache here to avoid N+1 query?

Comment (Nitish145) [reply app/serializers/api/v1/group_serializer.rb:7]: That would be great addition, thanks for this suggestion!!

**Comment by Nitish145** [comment]
@tachyons please check!

**Comment by Nitish145** [comment]
@tachyons can this be merged ? or some more changes required here..

**Comment by Nitish145** [comment]
@tachyons can we merge this now ?

**Comment by Nitish145** [comment]
Just to mention, CircleCI builds are failing on system specs.. @tachyons 

```Capybara::ElementNotFound:
  Unable to find visible link or button "Start Working"```

**Comment by Nitish145** [comment]
@tachyons any more changes here ?

**Pull request merged**